### PR TITLE
Added support for RankingMetrics with CrossValSummaryRunner

### LINKED ISSUE #5381: AutoML Ranking Error: "RankingMetrics Not Implemented"
When running the Ranking AutoML experiment on [this data](https://aka.ms/mlnet-resources/benchmarks/MSLRWeb10KTrain720kRows.tsv) from the Ranking sample, I seem to get a `Metric Microsoft.ML.Data.RankingMetrics not implemented` error during the `Execute` method.

I'm probably doing something wrong but I'm not sure what I'm missing.

```csharp
var context = new MLContext();

var data = context.Data.LoadFromTextFile<RankingData>("./ranking.tsv", separatorChar: '\t');

var settings = new RankingExperimentSettings
{
      MaxExperimentTimeInSeconds = 300,
      OptimizingMetric = RankingMetric.Ndcg
};

var experiment = context.Auto().CreateRankingExperiment(settings);

var results = experiment.Execute(data);
```

Here's the [full code](https://github.com/jwood803/MLNetExamples/blob/master/MLNetExamples/AutoMLRanking/Program.cs), if needed.

### PR DESCRIPTION
**Pull request by mstfbl**
Fix #5381 

Previously, `AutoFitRankingTest` was never testing `RankingMetrics` with `CrossValSummaryRunner` due to the row limit for cross validation testing, and as the MLSR dataset that `trainDataView` is pulling from is large:

https://github.com/dotnet/machinelearning/blob/f87a3bbd8adb12934dae0a0060813ce9b7500664/test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs#L123-L141

https://github.com/dotnet/machinelearning/blob/6bae29fc342bf192a36a69484d62db8d6266f8df/src/Microsoft.ML.AutoML/API/ExperimentBase.cs#L109-L130

As a result, this lacking of support for `RankingMetrics` with `CrossValSummaryRunner` was not noticed until Issue #5381. This PR adds support for this case, and edits the `RankingMetrics` AutoML unit tests to include testing of `RankingMetrics` with `CrossValSummaryRunner` as well.

### CODE REVIEW

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 1]
```diff
@@ -164,8 +179,8 @@ public void AutoFitRankingTest()
                 RunDetail<RankingMetrics> bestRun = experimentResults[i].BestRun;
                 Assert.True(experimentResults[i].RunDetails.Count() > 0);
                 Assert.NotNull(bestRun.ValidationMetrics);
-                Assert.True(experimentResults[i].RunDetails.Max(i => i.ValidationMetrics.NormalizedDiscountedCumulativeGains.Max() > .4));
-                Assert.True(experimentResults[i].RunDetails.Max(i => i.ValidationMetrics.DiscountedCumulativeGains.Max() > 20));
+                Assert.True(experimentResults[i].RunDetails.Max(i => i.ValidationMetrics.NormalizedDiscountedCumulativeGains.Max() > maxNdcg));
+                Assert.True(experimentResults[i].RunDetails.Max(i => i.ValidationMetrics.DiscountedCumulativeGains.Max() > maxDcg));
```
Comment (justinormont) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:183]: Currently I think you're checking the maximum value of NDCG@N for any `@N`, and for all created models. Instead I think we should check a specific `@N`. 

I didn't review any of the ranking additions to AutoML, is there an input for which `@N` to optimize? If so, I'd check that one.

If we don't know which N the user wants to optimize, I'd check the largest `@N` in NDCG@N:
`bestRun.ValidationMetrics.NormalizedDiscountedCumulativeGains.Last() > 0.6` instead of `Max()`.

This also only checks the one best model (ignoring the others -- removing the fori loop), and checks the NDCG@N for the largest `@N` specified.

I'm assuming the last array element is the NDCG@N and not the first; otherwise it'd be `bestRun.ValidationMetrics.NormalizedDiscountedCumulativeGains[0]`.

**Why check just the one `@N` value?**
We would expect as the `@N` increases, the metric should increase too. I'm not sure it's good to compare against a _single_ value for all `@N` values. This will allow a tighter bound on the expected value.

**Why check just the best model?**
In random sweeping there could easily be a very bad model, for instance if `NumTrees` get set to 2, it may make a poor model.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 2]
```diff
@@ -122,6 +122,19 @@ public void AutoFitRegressionTest()
...
+            // CrossValSummaryRunner is lower due to the 10 cross validation folds
+            // and due to the fact that CrossValSummaryRunner is used only when
+            // fewer than <1500 rows of training data is given.
+            AutoFitRankingTestTemplate(true, 0.4, 20);
+            AutoFitRankingTestTemplate(false, 0.13, 9.7);
+            AutoFitRankingCVTestTemplate(true, 0.4, 19);
+            AutoFitRankingCVTestTemplate(false, 0.13, 9.7);
+        }
+
+        private void AutoFitRankingTestTemplate(bool isUsingTrainValidateRunner, double maxNdcg, double maxDcg)
```
Comment (justinormont) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:137]: I would use another names besides `maxDcg` as this a term used within the calculation of the NDCG --  `NDCG = DCG / MaxDCG` ([code](https://github.com/dotnet/machinelearning/blob/d849ba4c7831586018d5b2cca6bdd2b1fa228ddf/src/Microsoft.ML.FastTree/Training/DcgCalculator.cs#L293)).

```suggestion
        private void AutoFitRankingTestTemplate(bool isUsingTrainValidateRunner, double minExpectedNdcg, double minExpectedDcg)
```

File: src/Microsoft.ML.AutoML/Experiment/Runners/CrossValSummaryRunner.cs [hunk 3]
```diff
@@ -144,9 +144,28 @@ private static TMetrics GetAverageMetrics(IEnumerable<TMetrics> metrics, TMetric
...
+                    dcg: GetAverageOfNonNaNScoresInNestedEnumerable(newMetrics.Select(x => x.DiscountedCumulativeGains)),
+                    ndcg: GetAverageOfNonNaNScoresInNestedEnumerable(newMetrics.Select(x => x.NormalizedDiscountedCumulativeGains)));
+                return result as TMetrics;
+            }
+
             throw new NotImplementedException($"Metric {typeof(TMetrics)} not implemented");
         }
 
+        private static double[] GetAverageOfNonNaNScoresInNestedEnumerable(IEnumerable<IEnumerable <double>> results)
+        {
```
Comment (justinormont) [inline src/Microsoft.ML.AutoML/Experiment/Runners/CrossValSummaryRunner.cs:162]: May want an assert that all incoming IEnumerables are the same length.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 4]
```diff
@@ -122,6 +122,19 @@ public void AutoFitRegressionTest()
...
+        {
+            // Testing Ranking with both TrainValidateRunner and CrossValSummaryRunner
+            // with corresponding baseline metrics. The metrics of Ranking with
+            // CrossValSummaryRunner is lower due to the 10 cross validation folds
+            // and due to the fact that CrossValSummaryRunner is used only when
+            // fewer than <1500 rows of training data is given.
+            AutoFitRankingTestTemplate(true, 0.4, 20);
+            AutoFitRankingTestTemplate(false, 0.13, 9.7);
+            AutoFitRankingCVTestTemplate(true, 0.4, 19);
+            AutoFitRankingCVTestTemplate(false, 0.13, 9.7);
```
Comment (justinormont) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:134]: Is the drop in score between the TrainValidate & CV runner due to the smaller dataset? Might be worth a run without subsampling to see that the scores are pretty close between the two.

Comment (mstfbl) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:134]: The lower benchmark metrics is due to the lower provided accuracies in this line, where no validation data is being provided:
https://github.com/dotnet/machinelearning/blob/ef8e085a194ba6e8ca942ea464381d589acc37a0/test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs#L160
As CV depends on validation data, this exact experiment may not be applicable for when running with `CrossValSummaryRunner`. @justinormont thoughts?

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs (hunk 2 above)
Comment (justinormont) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: You may want as "min" instead of "max", since your comparison is checking the metric value is larger than the given. Hence the given value is the minimum allowed value.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs (hunk 4 above)
Comment (mstfbl) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: @justinormont The metric results in the experiments below are differing significantly. Specifically, the fist experiment has a lot lower metrics than the other experiments:

https://github.com/dotnet/machinelearning/blob/5370692f6d1e6f1f686596d530032b4226dc9e80/test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs#L143-L160

Do you know why this is happening? Also, do you know why we are enforcing `SamplingKeyColumnName` and `GroupIdColumnName` to have the same column names for ranking?

https://github.com/dotnet/machinelearning/blob/5370692f6d1e6f1f686596d530032b4226dc9e80/src/Microsoft.ML.AutoML/API/ExperimentBase.cs#L67-L89

Comment (Lynx1820) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: SamplingKeys ensure that samples that have the same SamplingKey get grouped together within the training set and the test set. Ranking needs to have samples that have the same GroupId to be grouped together, so you can't have a sample with groupId A be both in the training set and in the test set. That means we need to enforce the samplingkey to be the GroupId only for ranking CV experiments.

Comment (mstfbl) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: Gotcha, thanks Keren!

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 5]
```diff
@@ -63,7 +63,7 @@ internal DataOperationsCatalog(IHostEnvironment env)
         /// results.
         /// </summary>
         /// <typeparam name="TRow">The user-defined item type.</typeparam>
-        /// <param name="data">The enumerable data containing type <typeparamref name="TRow"/> to convert to an<see cref="IDataView"/>.</param>
+        /// <param name="data">The enumerable data containing type <typeparamref name="TRow"/> to convert to a <see cref="IDataView"/>.</param>
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs:66]: shouldn't it be "an IDataView"?

Comment (mstfbl) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs:66]: It should be "... to convert to **a** Microsoft.ML.IDataView" as the `<see cref="IDataView"/> is printed as "Microsoft.ML.IDataView".

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 6]
```diff
@@ -122,6 +122,19 @@ public void AutoFitRegressionTest()
...
+            // CrossValSummaryRunner is lower due to the 10 cross validation folds
+            // and due to the fact that CrossValSummaryRunner is used only when
+            // fewer than <1500 rows of training data is given.
+            AutoFitRankingTestTemplate(true, 0.4, 20);
+            AutoFitRankingTestTemplate(false, 0.4, 28);
+            AutoFitRankingCVTestTemplate(true, 0.4, 19);
+            AutoFitRankingCVTestTemplate(false, 0.31, 15);
+        }
+
+        private void AutoFitRankingTestTemplate(bool isUsingTrainValidateRunner, double minExpectedNdcg, double minExpectedDcg)
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:137]: The function name and what it does is confusing. My guess is this template should only run train/validation ranking experiment because there's another function (AutoFitRankingCVTestTemplate) which looks like should only run cross-validation training.

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 7]
```diff
@@ -135,6 +148,8 @@ public void AutoFitRankingTest()
             var trainDataView = reader.Load(new MultiFileSource(DatasetUtil.GetMLSRDataset()));
             var testDataView = mlContext.Data.TakeRows(trainDataView, 500);
             trainDataView = mlContext.Data.SkipRows(trainDataView, 500);
+            if (!isUsingTrainValidateRunner)
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:151]: Maybe we can remove this if condition so this function only run train-validation?

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 8]
```diff
@@ -187,6 +204,8 @@ public void AutoFitRankingCVTest()
             var reader = new TextLoader(mlContext, GetLoaderArgsRank(labelColumnName, groupIdColumnName,
                 featuresColumnVectorNameA, featuresColumnVectorNameB));
             var trainDataView = reader.Load(DatasetUtil.GetMLSRDataset());
+            if (!isUsingTrainValidateRunner)
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:207]: Maybe remove this if so this function only run cross validation?

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 5 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs:66]: Oh that's tricky... thanks for explanation

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs (hunk 6 above)
Comment (mstfbl) [reply test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs:137]: Agreed, thank you.

**Comment by LittleLittleCloud** [comment]
LGTM

**Review by LittleLittleCloud — APPROVED** [review]

**Review by harishsk — APPROVED** [review]
